src/mainboard/sifive/hifive/src/main.rs

```diff
@@ -74,6 +74,27 @@ pub extern "C" fn _start_boot_hart(_hart_id: usize, fdt_address: usize) -> ! {
     clk.pwrite(b"on", 0).unwrap();
     uart0.pwrite(b"Done\r\n", 0).unwrap();
 
+    // QEMU doesn't emulate the SPI controller interface
+    if !is_qemu() {
```

Comment:
my only slight concern here is shipping binaries in boards with this test in. Is there some way we can avoid that or is there magic I don't understand that ensures we avoid it?

Reply:
The `is_qemu()` test or the read from 0x3000_0000? There's another `is_qemu()` at line 105 so I figured that was okay. I can take out the read from 0x3000_0000.

Reply:
Regarding this is_qemu() check: In x86 world checking for QEMU is often done by inspecting certain flags in the CPUID return values (don’t remember which right now).

I know nothing about RISCV or how QEMU emulates it, maybe something like that is possible as well, where you don’t read from a memory location that might have unforeseen consequences on real hardware?

Reply:
[Right now Oreboot reads the MSEL register](https://github.com/oreboot/oreboot/blob/master/src/soc/sifive/fu540/src/lib.rs). Apparently QEMU sets that register to 0x297, which isn't possible on actual hardware. I can't find where QEMU does that, but it does seem to work. I'll also note that there's a big ol' TODO right next to the check that says there might be a better way.

Reply:
I have to admit that I did not see that the is_qemu test was local for the respective board that is emulated here. I thought this was a universal test for anything riscv. Apologies for the confusion.

Just in case a board-agnostic test is wanted sometime in the future:
I just took a quick browse through the ISA. Is QEMU setting something unique in the mimpid cpu Register that maybe differentiates between QEMU and real HW?

Reply:
Unfortunately, the mvendorid/marchid/mimpid weren't implemented at the time that the FU540 SoC was created, so the hardware returns zeros. I suppose that means QEMU could return nonzero values, the trick is which ones?

Reply:
I think this is OK for now, just wanted to get it out there.